### ISSUE TITLE
Consider adding SecureArgument to make the guard behavior explicit

### ISSUE THREAD

**Issue by girtsl** [safakgur/guard#36]
**Is your feature request related to a problem?**
I've never been a fan of [flag arguments](https://martinfowler.com/bliki/FlagArgument.html). Apart from violating the SRP it also makes the code harder to read and less maintainable. It's also easier to make a mistake by passing the wrong flag value, one must always inspect the method's signature.

**Describe the solution you'd like**
Add a new factory method `SecureArgument` that does exactly what it says on the tin. Remove the `secure` flag from the `Argument` factory method to follow the SRP.

Consider:
```csharp
Guard.Argument(() => value, true).NotDefault();
```
vs
```csharp
Guard.SecureArgument(() => value).NotDefault();
```
IMHO the intent is much clearer and just by looking at the code there is no more guessing what that `true` does.

I realize this is a breaking change and would only be considered for the next major version, if at all. Nevertheless, I've already implemented it and will submit a PR.

**Comment by safakgur** [comment]
Hi Girts,

I appreciate the readability benefits Martin Fowler mentions in the post you linked. It was a good read, so thank you for that. But I still have mixed feelings about applying this for secure arguments.

* I really don't see how the current flag violates single responsibility as the only responsibility of the `Argument` method is to create an `ArgumentInfo<T>`. It doesn't even check the value of `secure`, it just assigns it to the `Secure` property and returns the new instance, so the extension methods that will actually validate the argument can check it.

* I'm not very proud of choosing the word "secure" for a flag that indicates "do not include argument value in the exception messages of failed validations". Maybe we should have called it `privateValue` or `doNotIncludeValueInMessages`. Having a `SecureArgument` method makes it look like it returns a different kind of guarded argument - a "Secure" one instead of an "Insecure" one, making my bad choice of the word even more explicit.

* Flag arguments are everywhere. It's an extremely common pattern that's widely used in the .NET framework itself. This shouldn't be a reason to choose flag arguments when developing a new library, but it makes me question the worth of introducing a breaking change to stop using flag arguments.

I want to stress that I *do* find your proposal more readable. I just don't see how the existing approach violates single responsibility. I also prefer not to keep the word, "secure" in upcoming versions.

I'm leaving the issue open as we have some time before v2 - I'm open to continuing the discussion.

Thank you for the suggestion.

**Comment by girtsl** [comment]
Hi,

One could argue that it's the `ArgumentInfo<T>` class that violates SRP by having two different behaviors, controlled by the `secure` flag and the `Argument` method is simply a facade which does nothing to hide this design decision. That being said, flag arguments have their use and it's not always practical to have two methods or create an inheritance hierarchy, e.g. `RegularArgumentInfo<T>: ArgumentInfo<T>` and `SensitiveArgumentInfo<T>: ArgumentInfo<T>`. 
Perhaps the issue will become more apparent if/when there is a need for yet another behavior and another flag needs to be added (or the implementation refactored).  

You will notice that my proposal only addresses the facade to make the API subjectively more readable. 

The word "secure" is likely not the best choice but I'm not sure I can offer any better options that capture the intent and the change in behavior in a concise manner. Frankly, it's not even a feature I use today or imagine using in the future so I can't offer anything that would be derived from a use case. Perhaps `SensitiveArgument` fits?

E.g.
```csharp
Guard.SensitiveArgument(() => value).NotDefault();
```

I also played with the thought of something along the lines of
```csharp
Guard.Argument(() => value).NotDefault().Sensitive();
```
but I think it's not prominent enough and easy to confuse with a validation check.

**Comment by safakgur** [comment]
Thanks, sensitive is actually much more fitting than secure.

Now that you mention it, there are actually two flags in `ArgumentInfo`, the other being `Modified` . Although, atm it can only be specified by calling the constructor directly. So what would happen if we decide to add a new flag, `Important` (not that we will ever have that one) - Would we have separate boolean parameters as we have now, or a single flag enum like `ArgumentFeatures`, or methods like `SensitiveArgument`, `ImportantArgument` and `SensitiveImportantArgument`? Maybe even a combination like `Argument` and `SensitiveArgument` methods accepting an optional `important` flag...

Back to reality: I don't think we'll ever introduce new flags and `Modified` needs to be set only when writing new, modifier guards. So adding a `SensitiveArgument` and removing the relevant parameter from the `Argument` makes sense. You can change Secure to Sensitive in your PR if you want, I can do it later as well.

Thanks again for the suggestion and effort.

**Comment by girtsl** [comment]
When writing my previous comment I overlooked the fact that `Argument<T>` is a struct and not a class so inheritance hierarchy is not even an option. Changing it to class means heap allocation and a major overhaul for the library, all for no good reason.
Thankfully, the "right" solution is much simpler than that. The flags `modified` and `secure` actually make the code use different [strategies](https://en.wikipedia.org/wiki/Strategy_pattern) for certain things (it seems that in the case of `modified` it might not be as simple as that, though). In the case of the `secure`/`sensitive` flag there are currently two strategies for resolving the argument value, both taking the form of trivial map functions:
- [identity](https://en.wikipedia.org/wiki/Identity_function) function when the flag is false - normal behavior
- [constant](https://en.wikipedia.org/wiki/Constant_function) function that returns `default` when the flag is true - the "sensitive" behavior.

```csharp
private static class ValueMapStrategy<T>
{
	public static readonly Func<T, T> Identity = value => value;
	public static readonly Func<T, T> Sensitive = value => default!;
}
```

Other strategies such as `Important` where we could possibly want to have the value in uppercase if it's a string (can't think of any examples for other data types) can be easily implemented as yet another strategy.

Thinking this would be a trivial exercise I started to implement the ability to specify value mapper strategy (i.e. replace the `sensitive` flag with `Func<T, T> valueMapper`) but quickly realized that the `sensitive` flag, too, does a bit more than appears on the surface. Similar to the value mapper there would also be a need for a message strategy and it would make sense to pack both value and message mappers in a single strategy. This is still doable but would require more widespread changes and deviates too far from the original issue.

Renamed all instances of "secure" and derived terms to "sensitive" and similar terms.
